src/sinks/util/buffer/json.rs

```diff
@@ -0,0 +1,103 @@
+use crate::sinks::util::Batch;
+use serde_json::value::RawValue;
+
+type BoxedRawValue = Box<RawValue>;
+
+/// A `batch` implementation for storing an array of serialized
+/// json values.
+///
+/// To ensure that `push` does not panic pair it with the use of
+/// `JsonArray::encode` to produce the correct `Vec<u8>`.
+#[derive(Default, Debug)]
+pub struct JsonArrayBuffer {
+    buffer: Vec<BoxedRawValue>,
+    size: usize,
+}
+
+impl JsonArrayBuffer {
+    /// Encoding via this will ensure that pushing into this batch will
+    // not panic.
+    pub fn encode(value: impl serde::Serialize) -> Result<Vec<u8>, crate::Error> {
+        serde_json::to_vec(&value).map_err(Into::into)
+    }
+}
+
+impl Batch for JsonArrayBuffer {
+    type Input = Vec<u8>;
```

Comment:
This seems wrong. Why accept raw bytes as opposed to an event or String or `Value` or something like that?

Reply:
So we can't do this because for our use case we would need to support  generic `Serialize` types which is not possible with the current trait since `push` is not generic. This would also mean that we have to force all buffers to be generic over `Serialize` which I don't think is a good idea.

The other option is to make JsonArrayBuffer generic over some Serialize type which would then 1) only allow you to pass one concrete type to serialize 2) would force a box if you want more than one type 3) would make this more complex since you would need to somehow name the generic Serialize type in places where you are not directly dealing with it. Being type erased is nice here.

So I decided that the simpler route makes much more sense _if_ we provide a specific encode fn that you should pair this with. Since, we are the only consumers of this I though that this is a fine assumption that is _documented_ and is easy to use.

I do think we can do some more unchecked stuff but didnt think it was directly worth the cost.